### ISSUE TITLE
ERROR  Run npm install to download the correct Chromium revision (1056772)

### ISSUE THREAD

**Issue by auedbaki** [puppeteer/puppeteer#9192]
### Bug description

Steps to reproduce the problem:

1. Deploy instance on render.com cloud working fine on localhost

Getting ERROR  Run npm install to download the correct Chromium revision (1056772)

Oct 30 09:25:19 PM   ERROR  Run npm install to download the correct Chromium revision (1056772).
Oct 30 09:25:19 PM  
Oct 30 09:25:19 PM    at ChromeLauncher.resolveExecutablePath (node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:120:27)
Oct 30 09:25:19 PM    at ChromeLauncher.executablePath (node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:166:25)
Oct 30 09:25:19 PM    at ChromeLauncher.launch (node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:70:37)
Oct 30 09:25:19 PM    at async exports.convert_to_pdf (api/controllers/client.js:287:21)

### Puppeteer version

19.2.0

### Node.js version

10.15.0

### npm version

8.19.2

### What operating system are you seeing the problem on?

macOS

### Relevant log output

```Shell
Oct 30 09:25:19 PM   ERROR  Run npm install to download the correct Chromium revision (1056772).
Oct 30 09:25:19 PM  
Oct 30 09:25:19 PM    at ChromeLauncher.resolveExecutablePath (node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:120:27)
Oct 30 09:25:19 PM    at ChromeLauncher.executablePath (node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:166:25)
Oct 30 09:25:19 PM    at ChromeLauncher.launch (node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:70:37)
Oct 30 09:25:19 PM    at async exports.convert_to_pdf (api/controllers/client.js:287:21)
```

**Comment by gRoussac** [comment]
Hi, 

I have the same issue, adding a .puppeteerrc.cjs is crashing the whole angular app and /opt/render/.cache/puppeteer/chrome/linux-1056772/ folder is well present. Only downgrading puppeteer does the trick, it seems very specific to how render spin up Linux and makes an image of the build and because it works fine with docker files (https://github.com/gRoussac/evaluator/tree/master/docker)

**Comment by kfitzgerald** [comment]
Running into this as well, I suspect it has to do with the breaking change

```
use ~/.cache/puppeteer for browser downloads (https://github.com/puppeteer/puppeteer/pull/9095)
```

In my case, dependencies are installed using a privileged system account and the app is run using a different unprivileged account. I suspect that since the cache path is now user-dependent, this is a new problem.

**Comment by gRoussac** [comment]
On Render the folder /opt/render/.cache/puppeteer/chrome/linux-1056772/ exists after build BUT the application is then moved to /opt/render/project/src and furthermore the .cache folder is not copied. I see a others of .something but not .cache.

Seems to me part of the issue is actually in this render workflow. As the [doc](https://pptr.dev/guides/configuration/) is saying 
```
this can cause problems if puppeteer is packed during some build step and moved to a fresh location. The following configuration can solve this issue (reinstall puppeteer to take effect):
```
which means puppeteer needs to be re installed on the Render "Start Command" with for instance ``` 'npm install puppeteer && npm run serve'``` but then deploy is declared failing (while it is not) without specific logs. So downgrading to 18.

@kfitzgerald in this [dockerfile](https://github.com/gRoussac/evaluator/blob/master/docker/Dockerfile) I am installing dependencies as unprivileged user and run the app as unprivileged user, maybe you can try to chown the .cache folder  to the unprivileged user but then you need to install the chrome_sandbox as root.

**Comment by kfitzgerald** [comment]
Definitely seems less like a bug and more of a breaking change gotcha with a misleading error message. Puppeteer is seemingly doing [exactly what is intended](https://pptr.dev/guides/configuration#changing-the-default-cache-directory) but the error message says "hey run npm install cuz chrome isn't installed" when in reality it was, but the directories do not align. 

## Workaround
For me, I had to update the tooling around the app, ensuring that the environment vars are consistent when installing and running the app using the different users.

When installing:
```
PUPPETEER_CACHE_DIR=/opt/puppeteer npm ci
```
> The cache dir needs to exist and be writeable by the installing user and readable by the app user

When running (e.g. in the systemctl service, add the same env var so the app will look at the common directory, e.g.
```
[Unit]
Description=My App Name Here
After=network.target auditd.service

[Service]
Type=simple
User=myunprivilegedusernamehere
Environment=PUPPETEER_CACHE_DIR=/opt/puppeteer
Restart=on-failure
WorkingDirectory=/opt/my-app-directory
ExecStart=/usr/bin/node /opt/my-app-directory/index.js

[Install]
WantedBy=multi-user.target
```

I chose not to use the `.puppeteerrc.cjs` file option, since running the app in different environments (e.g. local vs test vs prod) would cause that environment logic to have to be programmatically set in that file, e.g. `if (production) cacheDirectory = '...' else cacheDirectory = ...`

## Solution
Since the breaking change is documented, I feel that the error message should potentially be updated to help shed light on this change a bit more. It's not clear that the app might be using a different puppeteer cache directory than what was used during installation. Adding a bit more text or a url to an issue in the error message might be a reasonable solution.

**Comment by kfitzgerald** [comment]
Related: https://github.com/puppeteer/puppeteer/issues/9162

**Comment by jrandolf** [comment]
@kfitzgerald @gRoussac You can add a CJS configuration file with a cache directory that points to __dirname. I am not sure how this would be different from prod/dev (given the fact <19 behaviour can be replicated with the aforementioned method). Reinstallation is only required if you installed before modifying your configuration.

**Comment by gRoussac** [comment]
@jrandolf I believe your missunderstood the issue. Adding a config file to point to the same folder does not change anything to the problem.

**Comment by gRoussac** [comment]
@auedbaki what seems to work on my simple angular app is to create a ```.cache``` folder at the root of the project 

and in Render set this Environment variable ```PUPPETEER_DOWNLOAD_PATH``` to ```/opt/render/project/src/.cache```

The actual issue is that during npm install the folder used by Puppeteer in resolveExecutablePath to download Chrome is ```/opt/render/.cache``` (Cypress also does that) but the app is booted by Render in ``` /opt/render/project/src/``` thus the error message  'Run npm install to download the correct Chromium revision' because it can not find ```.cache``` two levels higher.

**Comment by xcbaojian** [comment]
Have you solved your problem? I have the same problem as you

/root/workspace/pupperteer_WcJ8/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:120
                    throw new Error(`Run \`npm install\` to download the correct Chromium revision (${this.puppeteer.browserRevision}).`);
                          ^

Error: Run `npm install` to download the correct Chromium revision (1056772).
    at ChromeLauncher.resolveExecutablePath (/root/workspace/pupperteer_WcJ8/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:120:27)
    at ChromeLauncher.executablePath (/root/workspace/pupperteer_WcJ8/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:166:25)
    at ChromeLauncher.launch (/root/workspace/pupperteer_WcJ8/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:70:37)

**Comment by gRoussac** [comment]
@xcbaojian 

Thsi problem is specific to each host, Heroku, Render etc and how they deploy an application.  Please understand that op specified "Deploy instance on render.com cloud working fine on localhost" so yes solved on Render with an env var.

You issue is probably similar, tries to see if Puppeteer installs in /root/.cache and if either adding a config file pointing to /root/workspace/pupperteer_WcJ8/ or if an env var works arround the issue as stated above.

**Comment by CodeRevolutionPlugins** [comment]
Same issue here as @kfitzgerald, having really hard time to install Puppeteer on Ubuntu, my unprivileged account got the same error [Error: Run `npm install` to download the correct Chromium revision (1056772)]
Would be really nice to have a step by step guide on how to solve this issue, as many of the links pointed out are now dead, for example: [https://pptr.dev/guides/configuring-puppeteer](https://pptr.dev/guides/configuring-puppeteer)
Thanks for your help!

**Comment by otrajman** [comment]
I copy/pasted the example configuration here:
https://www.npmjs.com/package/puppeteer

Then removed and reinstalled npm. Works locally and remote deploy.

It may have made sense to keep a default based on the example configuration
for better compatibility or included the example configuration in the
release notes for a few revs. Not an issue once it's configured though.

**Comment by reyronald** [comment]
I had the same issue and wanted to stop by and describe the solution I ended up going with in case it can help someone else.

I was getting the error above when trying to use a Puppeteer based feature in a deployed environment which runs on a Docker container. I was not getting the error when running the app locally outside of Docker.

Specifying a local cache directory through configuration [as described in the documentation here](https://pptr.dev/guides/configuration/#changing-the-default-cache-directory) worked **locally** after `rm -rf node_modules` and running `npm i` again, but for some reason it did NOT work in my Docker container 🤷‍♂️. I made sure to `docker-compose build` a fresh empty image with no cache to guarantee that puppeteer was seeing the configuration file and using it to change its cache directory but it wasn't being picked up. The resulting container did not have the cache directory I added in the `.puppeteerrc.cjs` and always used the default of `<homedir>/.cache/puppeteer`.

At this point I had already wasted too much time trying to figure it out so I gave up and just copied the default cache directory from the base stage of the Docker container to the release stage. In other words, I added this last step to my `Dockerfile`

```dockerfile
COPY --from=extended /root/.cache/puppeteer /.cache/puppeteer
```

This solved the issue so I'm ok for now. Would like to know why the configuration file approach didn't work just out of curiosity but probably won't be able to get an answer to that.

**Comment by jrandolf** [comment]
@reyronald It seems your setup is incorrect. The following is a pseudo-Dockerfile that works:

```Dockerfile
COPY /your-app /wherever
RUN npm i
# Etc. etc.
ENTRYPOINT ...
```

If your cache directory really is set up as described in the documentation, this should work.

**Comment by reyronald** [comment]
@jrandolf I can assure you the cache directory is set up as in the documentation because like I said it works as expected when installing the package locally. I tried both the configuration file and the env var and made sure the config file was visible to the container when installing packages.

Didn't see anything else obvious in the dockerfile that could be interfering. There's no other package that could be overwriting this directory either.

You don't have to answer though, like I said I'm already unblocked. Just wanted to document that for other folks who might run into this.

**Comment by gRoussac** [comment]
A simple Env var should indeed work but

> Didn't see anything else obvious in the dockerfile that could be interfering. There's no other package that could be overwriting this directory either.
> 

I would recommend to use a normal user to npm install (then you need a sandbox). ```.cache``` should not be in /root/ it should be in a user directory and then try to see where you are when npm install.
```
ARG USERNAME=user
RUN useradd --user-group --system --create-home --no-log-init ${USERNAME}
WORKDIR /home/${USERNAME}
USER $USERNAME
RUN npm install --loglevel=error && npm run build --loglevel=error
RUN pwd && ls -al .cache/puppeteer/chrome/linux-1056772/chrome-linux/

# v 19.2.0 This is the sandboxing part, requires a chrome.json file
USER root
ARG INSTALL_DIR=.cache/puppeteer/chrome/linux-1056772/chrome-linux/
RUN cd  ${INSTALL_DIR} && chown root:root chrome_sandbox && chmod 4755 chrome_sandbox && cp -p chrome_sandbox /usr/local/sbin/chrome-devel-sandbox
```

**Comment by AdirBuskila** [comment]
Hi there, I've been getting the error from the post above trying to deploy on Render with Node.js


`Nov 9 08:58:06 PM   1. you did not perform an installation before running the script (e.g. `npm install`) or
Nov 9 08:58:06 PM   2. your cache path is incorrectly configured (which is: /opt/render/.cache/puppeteer).
Nov 9 08:58:06 PM  For (2), check out our guide on configuring puppeteer at https://pptr.dev/guides/configuration.Problem getting star data at server`

and this is my puppeteerrc.cjs

`const { join } = require('path');

/**
 * @type {import("puppeteer").Configuration}
 */
module.exports = {
  // Changes the cache location for Puppeteer.
  cacheDirectory: join(__dirname, '.cache', 'puppeteer'),
};
`

any solution for this problem?

**Comment by reyronald** [comment]
> @reyronald A simple Env var should indeed work but
> 
> > Didn't see anything else obvious in the dockerfile that could be interfering. There's no other package that could be overwriting this directory either.
> 
> I would recommend to use a normal user to npm install (then you need a sandbox). `.cache` should not be in /root/ it should be in a user directory and then try to see where you are when npm install.
> 
> ```
> ARG USERNAME=user
> RUN useradd --user-group --system --create-home --no-log-init ${USERNAME}
> WORKDIR /home/${USERNAME}
> USER $USERNAME
> RUN npm install --loglevel=error && npm run build --loglevel=error
> RUN pwd && ls -al .cache/puppeteer/chrome/linux-1056772/chrome-linux/
> 
> # v 19.2.0 This is the sandboxing part, requires a chrome.json file
> USER root
> ARG INSTALL_DIR=.cache/puppeteer/chrome/linux-1056772/chrome-linux/
> RUN cd  ${INSTALL_DIR} && chown root:root chrome_sandbox && chmod 4755 chrome_sandbox && cp -p chrome_sandbox /usr/local/sbin/chrome-devel-sandbox
> ```

Are you saying that it could be a permission issue? As a user I would expect the installation to error with a meaningful permission related message instead of silently failing and defaulting to the home's cache directory, is that what happens in that case? @gRoussac

**Comment by gRoussac** [comment]
@AdirBuskila try in Render configuration to set this Environment variable ```PUPPETEER_DOWNLOAD_PATH``` to ```/opt/render/project/src/.cache```

@reyronald it is not a permission issue, it's a location issue.If you install as root, Puppeteer downloads Chrome in ```/root/.cache/puppeteer``` during npm install, that is why you need to copy it aftwerwards.

**Comment by reyronald** [comment]
> @reyronald it is not a permission issue, it's a location issue.If you install as root, Puppeteer downloads Chrome in /root/.cache/puppeteer during npm install, that is why you need to copy it aftwerwards.

But why is it installing it there if I'm specifying a different directory to install in with `.puppeteerrc.cjs`. The directory I was configuring it wasn't even named `.cache`  but it still ended up in `/root/.cache`

**Comment by AdirBuskila** [comment]
@gRoussac I did as you said and unfortunately I got the same error

**Comment by gRoussac** [comment]
@reyronald 
> But why is it installing it there if I'm specifying a different directory to install in with `.puppeteerrc.cjs`. The directory I was configuring it wasn't even named `.cache` but it still ended up in `/root/.cache`

I can not answer that question,.puppeteerrc.cjs didn't work for my case, imho there is a bug somewhere, when I tried a simple docker file to see the diffs with a Render install, something I noticed is that it was better to have Chrome being installed in ```/home/user/.cache/puppeteer``` and run the app with this user (moreover I don't want to fetch untrusted sites as root). Then runtime would find /home/user/.cache/puppeteer.

@AdirBuskila 
Either try to change ```puppeteerrc.cjs``` to ``` join(__dirname, 'project/src/' ,'.cache', 'puppeteer')```
or please try again, remove puppeteerrc.cjs, set the env var in Render (maybe creating a .cache folder at the root of your node app would help but I doubt) and npm install. The error should not be the same ``` /opt/render/.cache/puppeteer``` is two levels higher than your app is. If you need to see where are some folders in Render logs, you can use classic linux commands like ```pwd && ls -al .cache && npm start``` for instance in start command.

**Comment by AdirBuskila** [comment]
@gRoussac 
I changed the `puppeteerrc.cjs` and still no luck
any chance you'll look at my repo and tell me what I'm missing?
it's a very small project and that puppeteer error is very frustrating 
repo: https://github.com/AdirBuskila/star-info

**Comment by otrajman** [comment]
I believe the file needs to be ".puppeteerrc.cjs" with a leading dot.

**Comment by lowzyyy** [comment]
To add to this problem @AdirBuskila. For build i have to type this:

`npm install && node node_modules/puppeteer/install.js`

and my `.puppeteerrc.cjs` :
const { join } = require("path");

/**
 * @type {import("puppeteer").Configuration}
 */
module.exports = {
  // Changes the cache location for Puppeteer.
  cacheDirectory: join(__dirname, ".cache", "puppeteer"),
};

If you add `'project/src/'` to join it will add unnececery project/src because its already included in the path

**Comment by elioenays** [comment]
I resolved the issue by installing puppeteer again after running `npm install` in my build command.
i'm using this build command
`npm i && npx prisma migrate deploy && npx prisma db seed && npm install puppeteer && npm run build`

and I have the .puppeteerrc.cjs file with the cache configuration recommended by the documentation at https://www.npmjs.com/package/puppeteer

```javascript
const { join } = require('path')

/**
 * @type {import("puppeteer").Configuration}
 */
module.exports = {
  // Changes the cache location for Puppeteer.
  cacheDirectory: join(__dirname, '.cache', 'puppeteer'),
}
```

**Comment by estatistics** [comment]
SOLVED (debian linux 12)
path was /home/USER/.cache/puppeteer/chrome/linux-117.0.5938.92/chrome-linux64/
path copied as /home/USER/.cache/puppeteer/chrome/linux-1108766/chrome-linux/

Please add in script to print the expected path that puppeeter is searching chromium.

note that running as sudo yarn PROGRAM is searching in root/.cache while yarn PROGRAM in /home/user...
but the latest was throwing errors

So, `sudo npm run build --loglevel=error` or `sudo yarn run build --loglevel=error` and then `yarn start` or `npm start`
